src/symbolize/mod.rs

```diff
@@ -108,26 +111,82 @@ impl fmt::Debug for Symbol {
     }
 }
 
+// Maybe a parsed C++ symbol, if parsing the mangled symbol as Rust failed.
+#[cfg(feature = "cpp_demangle")]
+struct OptionCppSymbol<'a>(Option<::cpp_demangle::BorrowedSymbol<'a>>);
+
+#[cfg(feature = "cpp_demangle")]
+impl<'a> OptionCppSymbol<'a> {
+    fn parse(input: &'a [u8]) -> OptionCppSymbol<'a> {
+        OptionCppSymbol(::cpp_demangle::BorrowedSymbol::with_tail(input)
+                            .ok()
+                            .map(|(sym, _tail)| sym))
+    }
+
+    fn none() -> OptionCppSymbol<'a> {
+        OptionCppSymbol(None)
+    }
+
+    fn as_str(&self) -> Option<Cow<'a, str>> {
+        self.0.as_ref().map(|cpp| Cow::from(format!("{}", cpp)))
+    }
+}
+
+// Make sure to keep this zero-sized, so that the `cpp_demangle` feature has no
+// cost when disabled.
+#[cfg(not(feature = "cpp_demangle"))]
+struct OptionCppSymbol<'a>(PhantomData<&'a ()>);
+
+#[cfg(not(feature = "cpp_demangle"))]
+impl<'a> OptionCppSymbol<'a> {
+    fn parse(_: &'a [u8]) -> OptionCppSymbol<'a> {
+        OptionCppSymbol(PhantomData)
+    }
+
+    fn none() -> OptionCppSymbol<'a> {
+        OptionCppSymbol(PhantomData)
+    }
+
+    fn as_str(&self) -> Option<Cow<'a, str>> {
+        None
+    }
+}
+
 /// A wrapper around a symbol name to provide ergonomic accessors to the
 /// demangled name, the raw bytes, the raw string, etc.
 pub struct SymbolName<'a> {
     bytes: &'a [u8],
     demangled: Option<Demangle<'a>>,
+    cpp_demangled: OptionCppSymbol<'a>,
 }
 
 impl<'a> SymbolName<'a> {
     /// Creates a new symbol name from the raw underlying bytes.
     pub fn new(bytes: &'a [u8]) -> SymbolName<'a> {
         let demangled = str::from_utf8(bytes).ok().map(demangle);
+
+        // Only try and parse a C++ symbol if we didn't parse a Rust symbol.
+        let cpp = if demangled.is_none() {
+            OptionCppSymbol::parse(bytes)
+        } else {
+            OptionCppSymbol::none()
+        };
+
         SymbolName {
             bytes: bytes,
             demangled: demangled,
+            cpp_demangled: cpp,
         }
     }
 
-    /// Returns the raw symbol name as `&str` if the symbols is valid utf-8.
-    pub fn as_str(&self) -> Option<&'a str> {
-        self.demangled.as_ref().map(|s| s.as_str())
+    /// Returns the raw symbol name as a `str` if the symbols is valid utf-8.
+    pub fn as_str(&self) -> Option<Cow<'a, str>> {
```

Comment:
Could this stay as it was but leave a TODO to fix this on the next major version change?